**Step 1: Data bilan tanishish**
   - CSV faylni ochib, birinchi 10 qatorni ko‘rish. 
   - Column nomlari va sonini tekshirish. 
   - Har bir columndagi missing values sonini aniqlash. 
   - Data turlari (string, numeric, date, json) haqida tushuncha hosil qilish. 

In [1]:
import pandas as pd
# CSV faylni ochish
df=pd.read_csv(r'C:\Users\qaleb\Downloads\Telegram Desktop\super_dirty_students.csv')
# birinchi 10 qatorni ko`rish`
df.head(10) 


,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20,NaN,NaN,+1-619-379-4152x102,Katieland,someonegmail.com,1662247364,Data Science,NaN,active,3.72,good,$135,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20,Female,ninety,NaN,Dawnburgh,psmith@chen.com,2017/08/29,Data Science,NaN,active,1.88,excellent,$152,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20,NaN,ninety,NaN,Lake Stevenmouth,NaN,2017-08-14,DATA SCIENCE,NaN,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,twenty,fmale,81,NaN,Port Pamelafort,special,1973-09-17,data-sciens,NaN,inactive,0,good,175 USD,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20,Female,NaN,001-182-659-5631x02803,Kingburgh,six,2023/11/25,data_sciense,NaN,pending,4.37,average,$282,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
5,6,David Martinez,20,Male,ninety,001-074-828-6016x937,Hugheshaven,nathangibbs@hotmail.com,1603493647,DATA SCIENCE,110,pending,15.5,average,"85,00",1983-04-30T01:46:36,"Apartment 42, North Jamesfurt district, Tashke...","{hobbies:['order', 'keep']}"
6,7,NaN,18 years,FEMALE,71,314.018.0928x3134,Sharpfurt,nguyenlisa@@horn.org,1982-01-16,data science,NaN,active,0,excellent,275.0,19/11/2004 11:42 PM,"Lake Tracishire 29-kv, dom 8, Tashkent","{'hobbies': ['fall', 'teacher'], 'skills': {'t..."
7,8,Nicholas Dennis,21,FEMALE,ninety,8091556011,Matthewfort,avoid,2000.06.26,python,63,active,four point five,excellent,135,1978/09/23,"70576 Butler Harbor Suite 847, Janetberg",INVALID_JSON_DATA
8,9,NaN,twenty,fmale,77,854-964-5423,South Paulchester,stacysmith@taylor.biz,2000-03-15,ds,110%,active,15.5,excellent,104 USD,1995-10-20,"Williamsborough 2-kv, dom 18, Tashkent","{'hobbies': ['require', 'open'], 'skills': {'t..."
9,10,Elizabeth Villegas,22,femlae,NaN,(205)467-4476x0583,South Dianatown,anthonywalker@@gmail.com,2013-02-11T07:12:29,data_sciense,99,inactive,"3,7",average,"215,00",2014-12-16T06:44:51,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['management', 'to'], 'skills': {'..."


In [2]:
#  columns
df.columns

Index(['student_id', 'name', 'age', 'gender', 'score', 'phone', 'city',
       'email', 'date_of_join', 'course', 'attendance', 'status', 'gpa',
       'remarks', 'money_spent', 'event_time', 'address_raw', 'profile_json'],
      dtype='str')

In [3]:
df.shape

(1000, 18)

In [ ]:
#data frame haqida umumiy malumot
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   student_id    1000 non-null   int64
 1   name          665 non-null    str  
 2   age           892 non-null    str  
 3   gender        781 non-null    str  
 4   score         754 non-null    str  
 5   phone         629 non-null    str  
 6   city          1000 non-null   str  
 7   email         849 non-null    str  
 8   date_of_join  1000 non-null   str  
 9   course        1000 non-null   str  
 10  attendance    674 non-null    str  
 11  status        1000 non-null   str  
 12  gpa           859 non-null    str  
 13  remarks       796 non-null    str  
 14  money_spent   1000 non-null   str  
 15  event_time    1000 non-null   str  
 16  address_raw   1000 non-null   str  
 17  profile_json  1000 non-null   str  
dtypes: int64(1), str(17)
memory usage: 476.6 KB


In [3]:
# missing values number per column
df.isna().sum()

student_id        0
name            335
age             108
gender          219
score           246
phone           371
city              0
email           151
date_of_join      0
course            0
attendance      326
status            0
gpa             141
remarks         204
money_spent       0
event_time        0
address_raw       0
profile_json      0
dtype: int64

In [ ]:
#data typelar haqida malumot
df.dtypes

student_id      int64
name              str
age               str
gender            str
score             str
phone             str
city              str
email             str
date_of_join      str
course            str
attendance        str
status            str
gpa               str
remarks           str
money_spent       str
event_time        str
address_raw       str
profile_json      str
dtype: object

**Step 2: Oddiy string tozalash**
   - Har bir string columnning bo‘sh joylarini olib tashlash (strip). 
   - Bo‘sh stringlarni NaN yoki Null bilan almashtirish. 

In [3]:
string_columns=df.select_dtypes(include='str').columns #str columnni topish
df[string_columns]=df[string_columns].apply(lambda x: x.str.strip()) #bo`sh joylarni olib tashlash`
(string_columns=="").sum() #bo`sh str borligini aniqlash



np.int64(0)

In [4]:
# bo`sh strni NaN ga almashtirish
import numpy as np
df.replace("", np.nan, inplace=True)


,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20,NaN,NaN,+1-619-379-4152x102,Katieland,someonegmail.com,1662247364,Data Science,NaN,active,3.72,good,$135,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20,Female,ninety,NaN,Dawnburgh,psmith@chen.com,2017/08/29,Data Science,NaN,active,1.88,excellent,$152,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20,NaN,ninety,NaN,Lake Stevenmouth,NaN,2017-08-14,DATA SCIENCE,NaN,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,twenty,fmale,81,NaN,Port Pamelafort,special,1973-09-17,data-sciens,NaN,inactive,0,good,175 USD,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20,Female,NaN,001-182-659-5631x02803,Kingburgh,six,2023/11/25,data_sciense,NaN,pending,4.37,average,$282,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992.11.13,DATA SCIENCE,NaN,inactive,NaN,excellent,$64,14/10/2007 06:53 PM,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20,NaN,NaN,001-205-065-8737,Rogerborough,katiepage@hensley.com,2024.10.05,DATA SCIENCE,NaN,pending,"3,7",NaN,143,2021-06-21T07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA
997,998,NaN,20,MALE,NaN,264-577-8585,South Mary,someonegmail.com,19/11/2021 02:23 AM,data_sciense,95,inactive,-2,excellent,79.0,1975/05/11,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21,FEMALE,86,001-663-118-1327x207,Port Rachael,richardgreen@@shannon-jenkins.info,1625723463,data_sciense,110%,active,3.06,average,"238,00",2015.05.24,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec..."


**Step 3: Raqam va sanalarni tozalash**
   - Age, score, GPA, attendance, money_spent kabi numeric columnlarni tozalash va int yoki float formatiga keltirish. 
   - Date columnlarni pandas datetime formatiga o‘tkazish. 

In [5]:
#harflar va keraksiz belgilarni tozalash
numeric_columns=['age', 'score', 'gpa', 'attendance', 'money_spent']
for col in numeric_columns:
    df[col]=(
        df[col]
        .astype(str)
        .str.replace(r"[^\d.\-]", "", regex=True)
    )
#data typeni numeric qilish
for col in numeric_columns:
   df[col]=pd.to_numeric(df[col], errors='coerce')
# intga, floatga o`tkazish
df['age'] = df['age'].astype('Int64')      
df['attendance'] = df['attendance'].astype('Int64')

df['gpa'] = df['gpa'].astype(float)
df['money_spent'] = df['money_spent'].astype(float)

df.dtypes

student_id        int64
name                str
age               Int64
gender              str
score           float64
phone               str
city                str
email               str
date_of_join        str
course              str
attendance        Int64
status              str
gpa             float64
remarks             str
money_spent     float64
event_time          str
address_raw         str
profile_json        str
dtype: object

In [6]:
# date columnlarni pandas datetimega o`tkazish`

df['date_of_join']=pd.to_datetime(df['date_of_join'], format='mixed', errors='coerce')
df['event_time']=pd.to_datetime(df['event_time'], format='mixed', errors='coerce')


df[['date_of_join', 'event_time']] 

,date_of_join,event_time
0,NaT,NaT
1,2017-08-29 00:00:00,2001-11-10 04:19:00
2,2017-08-14 00:00:00,NaT
3,1973-09-17 00:00:00,NaT
4,2023-11-25 00:00:00,2013-10-13 00:00:00
...,...,...
995,1992-11-13 00:00:00,2007-10-14 18:53:00
996,2024-10-05 00:00:00,2021-06-21 07:46:27
997,2021-11-19 02:23:00,1975-05-11 00:00:00
998,NaT,2015-05-24 00:00:00


**Step 4: Email va phone validation **
   - Emaillarni kichik harflarga o‘tkazish va noto‘g‘ri formatdagi emaillarni aniqlash. 
   - Telefon raqamlarni tozalash va yagona standart formatga keltirish (masalan: +998...). 

In [73]:
df['email'].head(30)

0                someonegmail.com
1                 psmith@chen.com
2                             NaN
3                         special
4                             six
5         nathangibbs@hotmail.com
6            nguyenlisa@@horn.org
7                           avoid
8           stacysmith@taylor.biz
9        anthonywalker@@gmail.com
10                          stand
11             annalee@@gmail.com
12               someonegmail.com
13     aphillips@holland-diaz.org
14           kevin65@maxwell.info
15                            NaN
16                       yourself
17                          sport
18                           down
19                          voice
20              preid@hotmail.com
21         rrodriguez@@newman.com
22                            NaN
23                            NaN
24        urodriguez@@ferrell.com
25                          about
26                      direction
27    benjamincollins@@davis.info
28          elizabeth49@yahoo.com
29            

In [7]:
# emaillarni kichik harfga o`tkazdik`
df['email']=df['email'].str.lower()

email_pattern = r'^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}$'
df['email_valid']=df['email'].str.match(email_pattern, na=False)
df['email_valid'].sum()


np.int64(286)

In [8]:
# invalid emaillarni nanga o`zgartirish`
df.loc[~df['email_valid'], 'email']=np.nan

df = df.drop('email_valid', axis=1)
df['email'].head(50)


0                                      NaN
1                          psmith@chen.com
2                                      NaN
3                                      NaN
4                                      NaN
5                  nathangibbs@hotmail.com
6                                      NaN
7                                      NaN
8                    stacysmith@taylor.biz
9                                      NaN
10                                     NaN
11                                     NaN
12                                     NaN
13              aphillips@holland-diaz.org
14                    kevin65@maxwell.info
15                                     NaN
16                                     NaN
17                                     NaN
18                                     NaN
19                                     NaN
20                       preid@hotmail.com
21                                     NaN
22                                     NaN
23         

In [9]:
# phone number tozalash

def standardize_phone(x):
    if pd.isna(x):
        return np.nan
    x=str(x)
    x=x.split('x')[0]
    x=''.join(filter(str.isdigit, x))

    if x.startswith('1') and len(x)==11:
        return '+'+x
    elif x.startswith('001') and len(x)>11:
        return '+' + x[2:]
    elif len(x)==10:
        return '+1'+x
    else:
        return np.nan
df['phone_clean']=df['phone'].apply(standardize_phone)
df['phone']=df['phone_clean']
df.drop('phone_clean', axis=1, inplace=True)
df['phone'].head(50)

0     +16193794152
1              NaN
2              NaN
3              NaN
4     +11826595631
5     +10748286016
6     +13140180928
7     +18091556011
8     +18549645423
9     +12054674476
10    +19868509629
11             NaN
12    +12235736311
13             NaN
14             NaN
15             NaN
16             NaN
17    +12425525722
18    +11790996883
19    +16692728730
20             NaN
21    +10388742999
22    +11697150937
23             NaN
24             NaN
25             NaN
26             NaN
27    +14042330292
28    +15211681029
29    +10468929598
30             NaN
31    +10548293587
32    +16180905816
33             NaN
34    +13565523035
35    +18930217782
36    +17731802562
37    +11386481737
38    +15266776949
39             NaN
40             NaN
41             NaN
42    +16981153243
43             NaN
44             NaN
45             NaN
46             NaN
47             NaN
48    +12718094741
49    +10108099177
Name: phone, dtype: str

**Step 5: JSON parsing**
   - JSON columnlarini ochish (profile_json). 
   - JSON ichidagi ma’lumotlarni alohida columnlarga ajratish: hobbies, skills, family, devices. 
   - JSON ichidagi array yoki nested objectlarni tekshirish va flatten qilish.

In [15]:
import ast 
import json
def json_parse(x):
    if pd.isna(x) or x=='':
        return {}
    try:
        return json.loads(x)
    except json.JSONDecodeError:
        try:
            return ast.literal_eval(x)
        except:
            return {}
df['profile_json_parsed'] = df['profile_json'].apply(json_parse)



In [16]:

import copy
df1=copy.deepcopy(df)

In [17]:
df1_flat = pd.json_normalize(
    df1['profile_json_parsed'],
    sep='_'
)
df1_flat

,hobbies,devices,skills_tech_python,skills_tech_excel,skills_tech_sql,skills_soft,family_siblings,family_income_father,family_income_mother
0,"[gun, nice]","[{'type': 'laptop', 'brand': 'HP', 'year': 202...",2.0,5.0,1.0,"[with, onto]",4.0,1198.0,1089.0
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,"[fast, clearly]","[{'type': 'laptop', 'brand': 'HP', 'year': 202...",5.0,3.0,1.0,"[despite, clearly]",1.0,263.0,202.0
3,"[left, role]","[{'type': 'laptop', 'brand': 'Lenovo', 'year':...",3.0,5.0,1.0,"[full, series]",4.0,508.0,698.0
4,"[sort, science]","[{'type': 'laptop', 'brand': 'Lenovo', 'year':...",1.0,4.0,1.0,"[whose, save]",3.0,862.0,1058.0
...,...,...,...,...,...,...,...,...,...
995,"[collection, his]","[{'type': 'laptop', 'brand': 'Dell', 'year': 2...",3.0,5.0,1.0,"[increase, everything]",5.0,1382.0,1335.0
996,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
997,"[street, read]","[{'type': 'laptop', 'brand': 'HP', 'year': 202...",1.0,4.0,4.0,"[dinner, while]",5.0,1285.0,1005.0
998,"[game, total]","[{'type': 'laptop', 'brand': 'Dell', 'year': 2...",2.0,2.0,1.0,"[present, cup]",3.0,541.0,571.0


In [18]:
df1_flat['hobbies']=df1_flat['hobbies'].apply(lambda x: ', '.join(x) if isinstance(x, list) else np.nan)
df1_flat['skills_soft']=df1_flat['skills_soft'].apply(lambda x: ', '.join(x) if isinstance(x, list) else np.nan)

In [19]:
import numpy as np

def extract_devices(devices):
    result = {
        'laptop_brand': np.nan,
        'laptop_year': np.nan,
        'phone_brand': np.nan,
        'phone_year': np.nan
    }
    
    if isinstance(devices, list):
        for d in devices:
            if d.get('type') == 'laptop':
                result['laptop_brand'] = d.get('brand')
                result['laptop_year'] = d.get('year')
            elif d.get('type') == 'phone':
                result['phone_brand'] = d.get('brand')
                result['phone_year'] = d.get('year')
                
    return pd.Series(result)

df1_flat[['laptop_brand','laptop_year','phone_brand','phone_year']] = \
    df1_flat['devices'].apply(extract_devices)


In [20]:
df1_flat

,hobbies,devices,skills_tech_python,skills_tech_excel,skills_tech_sql,skills_soft,family_siblings,family_income_father,family_income_mother,laptop_brand,laptop_year,phone_brand,phone_year
0,"gun, nice","[{'type': 'laptop', 'brand': 'HP', 'year': 202...",2.0,5.0,1.0,"with, onto",4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,"fast, clearly","[{'type': 'laptop', 'brand': 'HP', 'year': 202...",5.0,3.0,1.0,"despite, clearly",1.0,263.0,202.0,HP,2020.0,Samsung,2023.0
3,"left, role","[{'type': 'laptop', 'brand': 'Lenovo', 'year':...",3.0,5.0,1.0,"full, series",4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0
4,"sort, science","[{'type': 'laptop', 'brand': 'Lenovo', 'year':...",1.0,4.0,1.0,"whose, save",3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,"collection, his","[{'type': 'laptop', 'brand': 'Dell', 'year': 2...",3.0,5.0,1.0,"increase, everything",5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0
996,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
997,"street, read","[{'type': 'laptop', 'brand': 'HP', 'year': 202...",1.0,4.0,4.0,"dinner, while",5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0
998,"game, total","[{'type': 'laptop', 'brand': 'Dell', 'year': 2...",2.0,2.0,1.0,"present, cup",3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0


In [21]:
df1_flat.drop('devices', axis=1, inplace=True)

In [22]:
df

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json,profile_json_parsed
0,1,Claudia Short,20,NaN,NaN,+16193794152,Katieland,NaN,NaT,Data Science,<NA>,active,3.72,good,135.0,NaT,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20,Female,NaN,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,Data Science,<NA>,active,1.88,excellent,152.0,2001-11-10 04:19:00,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}",{}
2,3,Kathryn Moyer,20,NaN,NaN,NaN,Lake Stevenmouth,NaN,2017-08-14 00:00:00,DATA SCIENCE,<NA>,pending,NaN,excellent,185.0,NaT,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t...","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,<NA>,fmale,81.0,NaN,Port Pamelafort,NaN,1973-09-17 00:00:00,data-sciens,<NA>,inactive,0.00,good,175.0,NaT,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20,Female,NaN,+11826595631,Kingburgh,NaN,2023-11-25 00:00:00,data_sciense,<NA>,pending,4.37,average,282.0,2013-10-13 00:00:00,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t...","{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,DATA SCIENCE,<NA>,inactive,NaN,excellent,64.0,2007-10-14 18:53:00,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20,NaN,NaN,+12050658737,Rogerborough,katiepage@hensley.com,2024-10-05 00:00:00,DATA SCIENCE,<NA>,pending,37.00,NaN,143.0,2021-06-21 07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA,{}
997,998,NaN,20,MALE,NaN,+12645778585,South Mary,NaN,2021-11-19 02:23:00,data_sciense,95,inactive,-2.00,excellent,79.0,1975-05-11 00:00:00,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21,FEMALE,86.0,+16631181327,Port Rachael,NaN,NaT,data_sciense,110,active,3.06,average,23800.0,2015-05-24 00:00:00,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec...","{'hobbies': ['game', 'total'], 'skills': {'tec..."


In [23]:
df.drop('profile_json', axis=1, inplace=True)
df.drop('profile_json_parsed', axis=1, inplace=True)  
df.dtypes


student_id               int64
name                       str
age                      Int64
gender                     str
score                  float64
phone                      str
city                       str
email                      str
date_of_join    datetime64[us]
course                     str
attendance               Int64
status                     str
gpa                    float64
remarks                    str
money_spent            float64
event_time      datetime64[us]
address_raw                str
dtype: object

In [30]:
df=df.join(df1_flat)

In [31]:
df

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,skills_tech_excel,skills_tech_sql,skills_soft,family_siblings,family_income_father,family_income_mother,laptop_brand,laptop_year,phone_brand,phone_year
0,1,Claudia Short,20,NaN,NaN,+16193794152,Katieland,NaN,NaT,Data Science,...,5.0,1.0,"with, onto",4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0
1,2,NaN,20,Female,NaN,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,Data Science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,3,Kathryn Moyer,20,NaN,NaN,NaN,Lake Stevenmouth,NaN,2017-08-14 00:00:00,DATA SCIENCE,...,3.0,1.0,"despite, clearly",1.0,263.0,202.0,HP,2020.0,Samsung,2023.0
3,4,Ruben Wilson,<NA>,fmale,81.0,NaN,Port Pamelafort,NaN,1973-09-17 00:00:00,data-sciens,...,5.0,1.0,"full, series",4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0
4,5,Robert Pruitt,20,Female,NaN,+11826595631,Kingburgh,NaN,2023-11-25 00:00:00,data_sciense,...,4.0,1.0,"whose, save",3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,DATA SCIENCE,...,5.0,1.0,"increase, everything",5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0
996,997,Scott Clark,20,NaN,NaN,+12050658737,Rogerborough,katiepage@hensley.com,2024-10-05 00:00:00,DATA SCIENCE,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
997,998,NaN,20,MALE,NaN,+12645778585,South Mary,NaN,2021-11-19 02:23:00,data_sciense,...,4.0,4.0,"dinner, while",5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0
998,999,Brittany Barrett,21,FEMALE,86.0,+16631181327,Port Rachael,NaN,NaT,data_sciense,...,2.0,1.0,"present, cup",3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0


**Step 6: Address parsing**
  - Address columnni alohida qism va key-larga ajratish: shahar, tuman, pochta kodi. 
  - Raw address columnni saqlab, yangi columnlar hosil qilish (addr_city, addr_district, addr_postal). 

In [32]:
df[['address_raw']]
for addr in df['address_raw']:
    print(addr)


Apartment 37, South Kevin district, Tashkent, UZ, 100539
UZ 100332 Tashkent South Patricia
Wendyshire 12-kv, dom 1, Tashkent
Apartment 16, North Tamara district, Tashkent, UZ, 100097
BROKEN,ADDRESS,DATA,,,
Apartment 42, North Jamesfurt district, Tashkent, UZ, 100551
Lake Tracishire 29-kv, dom 8, Tashkent
70576 Butler Harbor Suite 847, Janetberg
Williamsborough 2-kv, dom 18, Tashkent
BROKEN,ADDRESS,DATA,,,
Apartment 34, North Samantha district, Tashkent, UZ, 100659
Apartment 13, Morrismouth district, Tashkent, UZ, 100315
Frankchester 9-kv, dom 20, Tashkent
UZ 100394 Tashkent Stevensborough
Davidborough 9-kv, dom 29, Tashkent
1965 Jacobs Branch Apt. 418, Timburgh
UZ 100399 Tashkent Joehaven
Apartment 22, Port Kevin district, Tashkent, UZ, 100106
BROKEN,ADDRESS,DATA,,,
Masontown 10-kv, dom 21, Tashkent
BROKEN,ADDRESS,DATA,,,
UZ 100063 Tashkent Lake Shaneland
0451 James Cliffs Suite 725, Wagnermouth
0258 Sean Keys Apt. 380, Kevinfort
06462 Buck Springs Apt. 169, North Lisa
Apartment 20, La

In [33]:
df1['addr']=df1['address_raw'].astype('str').apply(lambda x: x.split(', '))

In [34]:
df1['addr'].head(50)

0     [Apartment 37, South Kevin district, Tashkent,...
1                   [UZ 100332 Tashkent South Patricia]
2                   [Wendyshire 12-kv, dom 1, Tashkent]
3     [Apartment 16, North Tamara district, Tashkent...
4                              [BROKEN,ADDRESS,DATA,,,]
5     [Apartment 42, North Jamesfurt district, Tashk...
6              [Lake Tracishire 29-kv, dom 8, Tashkent]
7            [70576 Butler Harbor Suite 847, Janetberg]
8              [Williamsborough 2-kv, dom 18, Tashkent]
9                              [BROKEN,ADDRESS,DATA,,,]
10    [Apartment 34, North Samantha district, Tashke...
11    [Apartment 13, Morrismouth district, Tashkent,...
12                [Frankchester 9-kv, dom 20, Tashkent]
13                  [UZ 100394 Tashkent Stevensborough]
14                [Davidborough 9-kv, dom 29, Tashkent]
15              [1965 Jacobs Branch Apt. 418, Timburgh]
16                        [UZ 100399 Tashkent Joehaven]
17    [Apartment 22, Port Kevin district, Tashke

In [35]:
df1['addr_city']=df1['addr'].apply(lambda x: x[-3] if len(x)==5 else x[-1] if len(x)==3 else np.nan)


In [36]:
df1['addr_district']=df1['addr'].apply(lambda x: x[-4] if len(x)==5 else  np.nan)
df1['addr_postal']=df1['addr'].apply(lambda x: x[-1] if len(x)==5 and x[-1].isnumeric() else np.nan)
df1['addr_district'].isna().sum()

np.int64(797)

In [37]:
df['addr_city']=df1['addr_city']
df['addr_district']=df1['addr_district']
df['addr_postal']=df1['addr_postal']
df

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,family_siblings,family_income_father,family_income_mother,laptop_brand,laptop_year,phone_brand,phone_year,addr_city,addr_district,addr_postal
0,1,Claudia Short,20,NaN,NaN,+16193794152,Katieland,NaN,NaT,Data Science,...,4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0,Tashkent,South Kevin district,100539
1,2,NaN,20,Female,NaN,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,Data Science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,3,Kathryn Moyer,20,NaN,NaN,NaN,Lake Stevenmouth,NaN,2017-08-14 00:00:00,DATA SCIENCE,...,1.0,263.0,202.0,HP,2020.0,Samsung,2023.0,Tashkent,NaN,NaN
3,4,Ruben Wilson,<NA>,fmale,81.0,NaN,Port Pamelafort,NaN,1973-09-17 00:00:00,data-sciens,...,4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0,Tashkent,North Tamara district,100097
4,5,Robert Pruitt,20,Female,NaN,+11826595631,Kingburgh,NaN,2023-11-25 00:00:00,data_sciense,...,3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,DATA SCIENCE,...,5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0,Tashkent,Laurenhaven district,100254
996,997,Scott Clark,20,NaN,NaN,+12050658737,Rogerborough,katiepage@hensley.com,2024-10-05 00:00:00,DATA SCIENCE,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,NaN,NaN
997,998,NaN,20,MALE,NaN,+12645778585,South Mary,NaN,2021-11-19 02:23:00,data_sciense,...,5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0,Tashkent,New Brendan district,100394
998,999,Brittany Barrett,21,FEMALE,86.0,+16631181327,Port Rachael,NaN,NaT,data_sciense,...,3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0,NaN,NaN,NaN


**Step 7: Duplicate va missing data tekshirish**
   - Duplicate rowsni aniqlash va o‘chirish. 
   - Muhim columnlarda missing values mavjudligini aniqlash va qaror qabul qilish (fillna yoki dropna). 

In [38]:

df.duplicated(['city', 'course', 'status', 'money_spent']).sum() # dublicate yo`q

np.int64(0)

In [39]:
import numpy as np
df.replace("", np.nan, inplace=True)


,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,family_siblings,family_income_father,family_income_mother,laptop_brand,laptop_year,phone_brand,phone_year,addr_city,addr_district,addr_postal
0,1,Claudia Short,20,NaN,NaN,+16193794152,Katieland,NaN,NaT,Data Science,...,4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0,Tashkent,South Kevin district,100539
1,2,NaN,20,Female,NaN,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,Data Science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,3,Kathryn Moyer,20,NaN,NaN,NaN,Lake Stevenmouth,NaN,2017-08-14 00:00:00,DATA SCIENCE,...,1.0,263.0,202.0,HP,2020.0,Samsung,2023.0,Tashkent,NaN,NaN
3,4,Ruben Wilson,<NA>,fmale,81.0,NaN,Port Pamelafort,NaN,1973-09-17 00:00:00,data-sciens,...,4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0,Tashkent,North Tamara district,100097
4,5,Robert Pruitt,20,Female,NaN,+11826595631,Kingburgh,NaN,2023-11-25 00:00:00,data_sciense,...,3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,DATA SCIENCE,...,5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0,Tashkent,Laurenhaven district,100254
996,997,Scott Clark,20,NaN,NaN,+12050658737,Rogerborough,katiepage@hensley.com,2024-10-05 00:00:00,DATA SCIENCE,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,NaN,NaN
997,998,NaN,20,MALE,NaN,+12645778585,South Mary,NaN,2021-11-19 02:23:00,data_sciense,...,5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0,Tashkent,New Brendan district,100394
998,999,Brittany Barrett,21,FEMALE,86.0,+16631181327,Port Rachael,NaN,NaT,data_sciense,...,3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0,NaN,NaN,NaN


In [40]:
df.isna().sum()

student_id                0
name                    335
age                     226
gender                  219
score                   497
phone                   371
city                      0
email                   714
date_of_join            177
course                    0
attendance              326
status                    0
gpa                     299
remarks                 204
money_spent               0
event_time              161
address_raw               0
hobbies                 406
skills_tech_python      406
skills_tech_excel       406
skills_tech_sql         406
skills_soft             406
family_siblings         406
family_income_father    406
family_income_mother    406
laptop_brand            406
laptop_year             406
phone_brand             406
phone_year              406
addr_city               624
addr_district           797
addr_postal             797
dtype: int64

In [41]:
df_cleaned = df.dropna(how='all', subset=['score', 'hobbies', 'addr_district'], inplace=True)
df

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,family_siblings,family_income_father,family_income_mother,laptop_brand,laptop_year,phone_brand,phone_year,addr_city,addr_district,addr_postal
0,1,Claudia Short,20,NaN,NaN,+16193794152,Katieland,NaN,NaT,Data Science,...,4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0,Tashkent,South Kevin district,100539
2,3,Kathryn Moyer,20,NaN,NaN,NaN,Lake Stevenmouth,NaN,2017-08-14 00:00:00,DATA SCIENCE,...,1.0,263.0,202.0,HP,2020.0,Samsung,2023.0,Tashkent,NaN,NaN
3,4,Ruben Wilson,<NA>,fmale,81.0,NaN,Port Pamelafort,NaN,1973-09-17 00:00:00,data-sciens,...,4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0,Tashkent,North Tamara district,100097
4,5,Robert Pruitt,20,Female,NaN,+11826595631,Kingburgh,NaN,2023-11-25 00:00:00,data_sciense,...,3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0,NaN,NaN,NaN
5,6,David Martinez,20,Male,NaN,+10748286016,Hugheshaven,nathangibbs@hotmail.com,NaT,DATA SCIENCE,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,North Jamesfurt district,100551
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
994,995,Kelly Martin,<NA>,male,NaN,NaN,Teresaville,NaN,2005-02-26 23:42:53,Data Science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,Port Ashley district,100498
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,DATA SCIENCE,...,5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0,Tashkent,Laurenhaven district,100254
997,998,NaN,20,MALE,NaN,+12645778585,South Mary,NaN,2021-11-19 02:23:00,data_sciense,...,5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0,Tashkent,New Brendan district,100394
998,999,Brittany Barrett,21,FEMALE,86.0,+16631181327,Port Rachael,NaN,NaT,data_sciense,...,3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0,NaN,NaN,NaN


**Step 8: Data normalization**
  - Gender columnini standard formatga keltirish: Male / Female / Unknown. 
  - Course columnlarini yagona nomga keltirish: Data Science / Python / Other. 
  - Status columnlarini standart formatga keltirish (lower-case yoki uniform). 

In [42]:
import numpy as np

df['gender'] = (
    df['gender']
    .astype(str)                # hammasini stringga o'tkazamiz
    .str.strip()                # bo'shliqlarni olib tashlaymiz
    .str.lower()                # lower case
)

df['gender'] = np.where(
    df['gender'].str.startswith('f'), 'Female',
    np.where(
        df['gender'].str.startswith('m'), 'Male',
        'Unknown'
    )
)


In [43]:
df

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,family_siblings,family_income_father,family_income_mother,laptop_brand,laptop_year,phone_brand,phone_year,addr_city,addr_district,addr_postal
0,1,Claudia Short,20,Unknown,NaN,+16193794152,Katieland,NaN,NaT,Data Science,...,4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0,Tashkent,South Kevin district,100539
2,3,Kathryn Moyer,20,Unknown,NaN,NaN,Lake Stevenmouth,NaN,2017-08-14 00:00:00,DATA SCIENCE,...,1.0,263.0,202.0,HP,2020.0,Samsung,2023.0,Tashkent,NaN,NaN
3,4,Ruben Wilson,<NA>,Female,81.0,NaN,Port Pamelafort,NaN,1973-09-17 00:00:00,data-sciens,...,4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0,Tashkent,North Tamara district,100097
4,5,Robert Pruitt,20,Female,NaN,+11826595631,Kingburgh,NaN,2023-11-25 00:00:00,data_sciense,...,3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0,NaN,NaN,NaN
5,6,David Martinez,20,Male,NaN,+10748286016,Hugheshaven,nathangibbs@hotmail.com,NaT,DATA SCIENCE,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,North Jamesfurt district,100551
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
994,995,Kelly Martin,<NA>,Male,NaN,NaN,Teresaville,NaN,2005-02-26 23:42:53,Data Science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,Port Ashley district,100498
995,996,Gary Sawyer,22,Female,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,DATA SCIENCE,...,5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0,Tashkent,Laurenhaven district,100254
997,998,NaN,20,Male,NaN,+12645778585,South Mary,NaN,2021-11-19 02:23:00,data_sciense,...,5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0,Tashkent,New Brendan district,100394
998,999,Brittany Barrett,21,Female,86.0,+16631181327,Port Rachael,NaN,NaT,data_sciense,...,3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0,NaN,NaN,NaN


In [44]:
df['course']=(
    df['course']
    .astype(str)
    .str.strip()
    .str.lower()
)


In [45]:

df['course']=np.where(
    df['course'].str.startswith('d'), 'Data Science',
    np.where(
        df['course'].str.startswith('p'), 'Python',
        'Other'
    )
)

In [46]:
df['course'].head(50)

0     Data Science
2     Data Science
3     Data Science
4     Data Science
5     Data Science
6     Data Science
8     Data Science
9     Data Science
10    Data Science
11    Data Science
14          Python
15          Python
16    Data Science
17    Data Science
18    Data Science
19    Data Science
20    Data Science
21          Python
23    Data Science
24    Data Science
25    Data Science
26          Python
27          Python
28    Data Science
29    Data Science
31          Python
32    Data Science
33    Data Science
35          Python
36    Data Science
37    Data Science
39          Python
40    Data Science
41    Data Science
42    Data Science
43          Python
44    Data Science
45    Data Science
46    Data Science
47    Data Science
49    Data Science
50    Data Science
51    Data Science
52    Data Science
53    Data Science
54    Data Science
55          Python
57    Data Science
58    Data Science
60    Data Science
Name: course, dtype: str

In [47]:
df['status'].head(50)

0       active
2      pending
3     inactive
4      pending
5      pending
6       active
8       active
9     inactive
10      active
11    inactive
14     pending
15     pending
16      active
17    inactive
18      active
19     pending
20      active
21    inactive
23     pending
24     pending
25     pending
26    inactive
27    inactive
28    inactive
29     pending
31      active
32    inactive
33      active
35    inactive
36     pending
37      active
39    inactive
40      active
41    inactive
42     pending
43    inactive
44      active
45    inactive
46      active
47      active
49      active
50    inactive
51    inactive
52    inactive
53     pending
54    inactive
55    inactive
57     pending
58      active
60    inactive
Name: status, dtype: str

**Step 9: Final type conversion va export** 
  - Barcha columnlarning to‘g‘ri type da ekanligini tekshirish: string, int, float, datetime. 
  - Sanalarni yagona formatga keltirish (YYYY-MM-DD HH:MM:SS). 
  - Tozalangan data CSV faylga saqlash (super_dirty_students_cleaned.csv). 

In [48]:
df

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,family_siblings,family_income_father,family_income_mother,laptop_brand,laptop_year,phone_brand,phone_year,addr_city,addr_district,addr_postal
0,1,Claudia Short,20,Unknown,NaN,+16193794152,Katieland,NaN,NaT,Data Science,...,4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0,Tashkent,South Kevin district,100539
2,3,Kathryn Moyer,20,Unknown,NaN,NaN,Lake Stevenmouth,NaN,2017-08-14 00:00:00,Data Science,...,1.0,263.0,202.0,HP,2020.0,Samsung,2023.0,Tashkent,NaN,NaN
3,4,Ruben Wilson,<NA>,Female,81.0,NaN,Port Pamelafort,NaN,1973-09-17 00:00:00,Data Science,...,4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0,Tashkent,North Tamara district,100097
4,5,Robert Pruitt,20,Female,NaN,+11826595631,Kingburgh,NaN,2023-11-25 00:00:00,Data Science,...,3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0,NaN,NaN,NaN
5,6,David Martinez,20,Male,NaN,+10748286016,Hugheshaven,nathangibbs@hotmail.com,NaT,Data Science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,North Jamesfurt district,100551
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
994,995,Kelly Martin,<NA>,Male,NaN,NaN,Teresaville,NaN,2005-02-26 23:42:53,Data Science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,Port Ashley district,100498
995,996,Gary Sawyer,22,Female,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,Data Science,...,5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0,Tashkent,Laurenhaven district,100254
997,998,NaN,20,Male,NaN,+12645778585,South Mary,NaN,2021-11-19 02:23:00,Data Science,...,5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0,Tashkent,New Brendan district,100394
998,999,Brittany Barrett,21,Female,86.0,+16631181327,Port Rachael,NaN,NaT,Data Science,...,3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0,NaN,NaN,NaN


In [49]:
# saqlash
df.to_csv(
    "super_dirty_students_cleaned.csv",
    index=False,
    date_format="%Y-%m-%d %H:%M:%S"
)

Step 10: QA checks 
  - Original va cleaned row sonini solishtirish. 
   - Missing email va phone sonini tekshirish. 
  - Numeric columnlar (GPA, attendance, score) qiymatlari to‘g‘ri diapazonda ekanligini tekshirish. 
  - Duplicate rowlar yo‘qligini tasdiqlash. 

In [50]:
df.info()

<class 'pandas.DataFrame'>
Index: 835 entries, 0 to 999
Data columns (total 32 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   student_id            835 non-null    int64         
 1   name                  557 non-null    str           
 2   age                   646 non-null    Int64         
 3   gender                835 non-null    str           
 4   score                 503 non-null    float64       
 5   phone                 529 non-null    str           
 6   city                  835 non-null    str           
 7   email                 237 non-null    str           
 8   date_of_join          684 non-null    datetime64[us]
 9   course                835 non-null    str           
 10  attendance            563 non-null    Int64         
 11  status                835 non-null    str           
 12  gpa                   583 non-null    float64       
 13  remarks               662 non-null  

In [51]:
df[['email', 'phone']].isna().sum()
df.duplicated().sum()


np.int64(0)

In [52]:
df[['gpa', 'score', 'attendance']].head(50)

,gpa,score,attendance
0,3.72,NaN,<NA>
2,NaN,NaN,<NA>
3,0.00,81.0,<NA>
4,4.37,NaN,<NA>
5,15.50,NaN,110
6,0.00,71.0,<NA>
8,15.50,77.0,110
9,37.00,NaN,99
10,37.00,NaN,110
11,-2.00,71.0,110


In [1]:
# Numeric columnlarni float ga o'tkazib olamiz
numeric_cols = ["gpa", "attendance", "score"]

for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors="coerce")

# Range validation
invalid_gpa = df[(df["gpa"] < 0) | (df["gpa"] > 5)]
invalid_attendance = df[(df["attendance"] < 0) | (df["attendance"] > 100)]
invalid_score = df[(df["score"] < 0) | (df["score"] > 100)]

print("Invalid GPA rows:", len(invalid_gpa))
print("Invalid attendance rows:", len(invalid_attendance))
print("Invalid score rows:", len(invalid_score))


NameError: name 'pd' is not defined